File: tools/balance_60min_sim.mjs
/**
 * game02 簡易バランスシミュレーション（60分・バズ無し・ランダム期待値近似）
 */
const VIEWS_TICK = 5.0;
// シーンの initialViewsBaseMultiplier に相当（再調整用）
let INITIAL_VIEWS_MULT = 4.79;
const EARLY_RAND = (0.92 + 1.08) / 2;
const MID_RAND = (0.95 + 1.02) / 2;
const LATE_RAND = (0.70 + 1.0) / 2;
const POP_RAND = (0.95 + 1.05) / 2;

function computeSalePrice(base, mult, level) {
  level = Math.max(0, level);
  const raw = base * Math.pow(mult, level);
  const truncated = Math.floor(raw / 100) * 100;
  if (truncated < 1000) return 1;
  return truncated;
}

const SIDE_PANEL_SCALE = 0.329;

function streamingBase(n) {
  const nn = Math.max(0, n);
  return 15 + 105 * Math.pow(nn / 19, 1.4);
}

function streamingFinal(n, d) {
  const dd = Math.max(0, d);
  return streamingBase(n) * Math.pow(0.88, dd);
}

function targetViews(elapsed, p) {
  p = Math.max(0, p);
  if (elapsed <= 0) return 0;
  if (elapsed <= 100) {
    const t = elapsed / 100;
    return 0.35 * p * t;
  }
  if (elapsed <= 180) {
    const t = (elapsed - 100) / 80;
    return 0.35 * p + (0.5 * p - 0.35 * p) * t;
  }
  const lateSeconds = elapsed - 180;
  const lateTicks = lateSeconds / VIEWS_TICK;
  const lateBase = p / 100;
  return 0.5 * p + lateTicks * lateBase;
}

function stageRand(nextElapsed) {
  if (nextElapsed <= 100) return EARLY_RAND;
  if (nextElapsed <= 180) return MID_RAND;
  return LATE_RAND;
}

function moneyRate(pop) {
  if (pop < 10000) return 0.2;
  if (pop < 100000) return 0.4;
  if (pop < 1000000) return 0.7;
  return 1.0;
}

function applyUpgrade(st, name) {
  if (name === "ed01") st.ed01 = 1;
  else if (name === "ed02") st.ed02 = Math.min(3, st.ed02 + 1);
  else if (name === "ed03") st.ed03 = Math.min(5, st.ed03 + 1);
  else if (name === "ed04") st.ed04 = Math.min(5, st.ed04 + 1);
  else if (name === "ed05") st.ed05 = Math.min(3, st.ed05 + 1);
  else if (name === "ed51") st.ed51 = 1;
  else if (name === "st01") {
    st.T_stream += 1;
    st.talk_lv = Math.min(8, st.talk_lv + 1);
  } else if (name === "st02" || name === "st03") {
    st.T_stream += 1;
  } else if (name === "st04") {
    st.T_stream += 1;
    st.d_stream = Math.min(5, st.d_stream + 1);
  }
}

function buildPlan(mult) {
  const {
    ed02_bp, ed02_pm,
    ed03_bp, ed03_pm,
    ed04_bp, ed04_pm,
    ed05_bp, ed05_pm,
    ed51_bp, ed51_pm,
    st01_bp, st01_pm,
    st02_bp, st02_pm,
    st03_bp, st03_pm,
    st04_bp, st04_pm,
    ed01_price = 8000,
  } = mult;

  const events = [];
  const addStair = (tag, m1, m3, bp, pm, kind) => {
    const m2 = (m1 + m3) / 2;
    events.push([m1, kind, bp, pm, 0]);
    events.push([m2, kind, bp, pm, 1]);
    events.push([m3, kind, bp, pm, 2]);
  };

  events.push([2, "eb", 1000, 1, 0]);
  events.push([10, "eb", 1000, 1, 0]);
  events.push([30, "eb", 1000, 1, 0]);
  events.push([5, "ed01", ed01_price, 1, 0]);
  addStair("ed02", 7, 38, ed02_bp, ed02_pm, "ed02");
  addStair("ed03", 8, 39, ed03_bp, ed03_pm, "ed03");
  addStair("ed04", 9, 40, ed04_bp, ed04_pm, "ed04");
  addStair("ed05", 11, 45, ed05_bp, ed05_pm, "ed05");
  events.push([25, "ed51", ed51_bp, ed51_pm, 0]);
  addStair("st01", 4, 42, st01_bp, st01_pm, "st01");
  addStair("st02", 8, 28, st02_bp, st02_pm, "st02");
  addStair("st03", 9, 38, st03_bp, st03_pm, "st03");
  addStair("st04", 10, 50, st04_bp, st04_pm, "st04");

  events.sort((a, b) => a[0] - b[0]);

  const edLv = { ed02: 0, ed03: 0, ed04: 0, ed05: 0 };
  const stLv = { st01: 0, st02: 0, st03: 0, st04: 0 };
  const byMin = new Map();

  for (const [tmin, kind, bp, pm, lvOr0] of events) {
    const mn = Math.floor(tmin);
    if (!byMin.has(mn)) byMin.set(mn, []);
    if (kind === "eb") {
      byMin.get(mn).push(["editor", 1000]);
    } else if (kind === "ed01") {
      byMin.get(mn).push(["ed01", computeSalePrice(bp, pm, 0)]);
    } else if (kind === "ed51") {
      byMin.get(mn).push(["ed51", computeSalePrice(bp, pm, 0)]);
    } else if (edLv[kind] !== undefined) {
      const lv = edLv[kind];
      byMin.get(mn).push([kind, computeSalePrice(bp, pm, lv)]);
      edLv[kind]++;
    } else if (stLv[kind] !== undefined) {
      const lv = stLv[kind];
      byMin.get(mn).push([kind, computeSalePrice(bp, pm, lv)]);
      stLv[kind]++;
    }
  }

  return [...byMin.entries()].sort((a, b) => a[0] - b[0]);
}

function viewsTick(m, pOverride) {
  const p = Math.max(0, pOverride ?? m.p);
  let delta = 0;
  if (!m.initialDone) {
    let initial = Math.round(p * INITIAL_VIEWS_MULT);
    initial = Math.max(1, initial);
    m.initialDone = true;
    delta += initial;
  }
  const now = m.elapsed;
  const nxt = now + VIEWS_TICK;
  const baseDelta = Math.max(0, targetViews(nxt, p) - targetViews(now, p));
  const tickRaw = Math.max(0, baseDelta * stageRand(nxt));
  m.pendingFrac += tickRaw;
  let tickDelta = Math.floor(m.pendingFrac);
  if (tickDelta > 0) m.pendingFrac -= tickDelta;
  delta += tickDelta;
  const dpop = Math.round(delta * 0.2 * POP_RAND);
  m.elapsed = nxt;
  m.tickCount += 1;
  return { delta, dpop };
}

function runSim(plan, options = {}) {
  INITIAL_VIEWS_MULT = options.initialViewsMult ?? 4.79;
  const priceScale = options.priceScale ?? 1.0; // 1.0（スケールは buildPlan 側で反映済み）
  const st = {
    t: 0,
    pop: 100,
    money: 9900,
    minMoney: 9900,
    T_stream: 0,
    d_stream: 0,
    ed01: 0,
    ed02: 0,
    ed03: 0,
    ed04: 0,
    ed05: 0,
    ed51: 0,
    talk_lv: 0,
    total_money_in: 0,
    total_views: 0,
  };

  const spawnP = () => {
    const base = Math.max(0, st.pop);
    const adj = Math.round(base * (1 + 0.2 * Math.min(st.talk_lv, 8)));
    return Math.max(0, Math.round(adj * 1.0 * (1 + 0.2 * st.ed04)));
  };

  const streamSec = () => streamingFinal(Math.max(0, st.T_stream - st.d_stream), st.d_stream);
  const editSec = () => Math.max(1, Math.floor(120 * Math.pow(0.9, st.ed02)));
  const numEditors = () => {
    if (st.t < 120) return 0;
    if (st.t < 300) return 1;
    return 2;
  };

  const scheduleNext = (fromT) => {
    const s = streamSec();
    const e = editSec();
    const ne = numEditors();
    if (ne <= 0) return fromT + s + e;
    return fromT + Math.max(s, e / ne);
  };

  let movies = [{ p: spawnP(), elapsed: 0, pendingFrac: 0, pendingMoneyFrac: 0, pendingMonetized: 0, initialDone: false, tickCount: 0 }];
  let nextSpawn = scheduleNext(0);
  let purchaseIdx = 0;

  const maxSlots = () => 2 + st.ed05;

  const snapshots = [];
  for (let m = 0; m <= 60; m++) snapshots.push({ min: m, pop: 0, money: 0, totalIn: 0 });

  function spawnMovie() {
    const nm = { p: spawnP(), elapsed: 0, pendingFrac: 0, pendingMoneyFrac: 0, pendingMonetized: 0, initialDone: false, tickCount: 0 };
    movies.unshift(nm);
    while (movies.length > maxSlots()) movies.pop();
  }

  while (st.t < 3600 + 1e-6) {
    const minute = Math.floor(st.t / 60);
    while (purchaseIdx < plan.length && plan[purchaseIdx][0] <= minute) {
      const [, pairs] = plan[purchaseIdx];
      for (const [name, price] of pairs) {
        const scaled =
          name === "editor"
            ? price
            : Math.max(1, Math.floor(price * SIDE_PANEL_SCALE * priceScale));
        st.money -= scaled;
        st.minMoney = Math.min(st.minMoney, st.money);
        if (name !== "editor") applyUpgrade(st, name);
      }
      purchaseIdx++;
    }

    while (st.t + 1e-6 >= nextSpawn) {
      spawnMovie();
      nextSpawn = scheduleNext(nextSpawn);
    }

    let dpopTotal = 0;
    for (const m of movies) {
      const { delta, dpop } = viewsTick(m, m.p);
      if (delta > 0) {
        st.total_views += delta;
        m.pendingMonetized += delta;
      }
      dpopTotal += dpop;
      if (m.tickCount % 2 === 0) {
        const payable = Math.floor(m.pendingMonetized / 100);
        if (payable > 0) {
          const pv = payable * 100;
          m.pendingMonetized -= pv;
          const rate = moneyRate(m.p);
          const raw = pv * rate + m.pendingMoneyFrac;
          const dm = Math.floor(Math.max(0, raw));
          m.pendingMoneyFrac = Math.max(0, raw - dm);
          st.money += dm;
          st.total_money_in += dm;
          st.minMoney = Math.min(st.minMoney, st.money);
        }
      }
    }
    st.pop = Math.max(0, st.pop + dpopTotal);

    const snapMin = Math.floor(st.t / 60);
    if (snapMin <= 60 && snapshots[snapMin]) {
      snapshots[snapMin].pop = st.pop;
      snapshots[snapMin].money = st.money;
      snapshots[snapMin].totalIn = st.total_money_in;
    }

    st.t += VIEWS_TICK;
  }

  return { st, snapshots };
}

// --- main ---
const currentMult = {
  ed02_bp: 1000, ed02_pm: 1.35,
  ed03_bp: 1000, ed03_pm: 1.35,
  ed04_bp: 1000, ed04_pm: 1.32,
  ed05_bp: 1200, ed05_pm: 1.35,
  ed51_bp: 1500, ed51_pm: 1.0,
  st01_bp: 1000, st01_pm: 1.22,
  st02_bp: 1000, st02_pm: 1.22,
  st03_bp: 1200, st03_pm: 1.28,
  st04_bp: 1500, st04_pm: 1.25,
  ed01_price: 5000,
};

const plan = buildPlan(currentMult);

function findPriceScale(ivm) {
  let lo = 0;
  let hi = 1.0;
  for (let i = 0; i < 45; i++) {
    const mid = (lo + hi) / 2;
    const { st } = runSim(plan, { priceScale: mid, initialViewsMult: ivm });
    if (st.minMoney >= 0 && st.money >= 0) lo = mid;
    else hi = mid;
  }
  return lo;
}

const ivmTarget = 0.0188;
const ps = findPriceScale(ivmTarget);
const { st, snapshots } = runSim(plan, { priceScale: ps, initialViewsMult: ivmTarget });
let totalSpend = 0;
for (const [, pairs] of plan) {
  for (const [n, pr] of pairs) {
    totalSpend += n === "editor" ? pr : Math.max(1, Math.floor(pr * SIDE_PANEL_SCALE * ps));
  }
}
console.log("initialViewsMult", ivmTarget, "PRICE_SCALE", ps, "SCALED_TOTAL_SPEND", totalSpend);
console.log("FINAL pop", Math.round(st.pop), "minMoney", Math.round(st.minMoney), "money", Math.round(st.money));
console.log("SNAPSHOTS every 10m (cumulative_gross = 初期9900+累計売上):");
for (let m = 0; m <= 60; m += 10) {
  const s = snapshots[m];
  const cumulative = 9900 + s.totalIn;
  console.log(`${m}m pop=${Math.round(s.pop)} money=${Math.round(s.money)} cumulative_gross=${Math.round(cumulative)}`);
}

File: tools/balance_60min_sim.py
"""
game02 簡易バランスシミュレーション（60分・バズ無し・ランダム期待値近似）。
work_upload / WorkMovieUploadController の式に合わせた離散 5 秒ティック。
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import List, Tuple

# --- WorkMovie（コード準拠） ---
VIEWS_TICK = 5.0
MONEY_TICK = 10.0
INITIAL_VIEWS_MULT = 4.79
# ランダムは期待値で固定（仕様書用の再現性）
EARLY_RAND = (0.92 + 1.08) / 2
MID_RAND = (0.95 + 1.02) / 2
LATE_RAND = (0.70 + 1.00) / 2
POP_RAND = (0.95 + 1.05) / 2


def compute_sale_price(base: int, mult: float, level: int) -> int:
    level = max(0, level)
    raw = base * (mult**level)
    if raw < 0 or math.isnan(raw) or math.isinf(raw):
        return 1
    truncated = int(math.floor(raw / 100.0) * 100.0)
    if truncated < 1000:
        return 1
    return truncated


def streaming_base(n: int) -> float:
    nn = max(0, n)
    return 15.0 + 105.0 * ((nn / 19.0) ** 1.4)


def streaming_final(n: int, d: int) -> float:
    dd = max(0, d)
    return streaming_base(n) * (0.88**dd)


def target_views(elapsed: float, p: float) -> float:
    p = max(0.0, p)
    if elapsed <= 0:
        return 0.0
    if elapsed <= 100:
        t = elapsed / 100.0
        return 0.35 * p * t
    if elapsed <= 180:
        t = (elapsed - 100.0) / 80.0
        return (0.35 * p) + ((0.5 * p - 0.35 * p) * t)
    late_seconds = elapsed - 180.0
    late_ticks = late_seconds / VIEWS_TICK
    late_base = p / 100.0
    return 0.5 * p + (late_ticks * late_base)


def stage_rand(next_elapsed: float) -> float:
    if next_elapsed <= 100:
        return EARLY_RAND
    if next_elapsed <= 180:
        return MID_RAND
    return LATE_RAND


def money_rate(pop: int) -> float:
    if pop < 10_000:
        return 0.2
    if pop < 100_000:
        return 0.4
    if pop < 1_000_000:
        return 0.7
    return 1.0


@dataclass
class Movie:
    p: int
    elapsed: float = 0.0
    pending_frac: float = 0.0
    pending_money_frac: float = 0.0
    pending_monetized: int = 0
    initial_done: bool = False
    tick_count: int = 0
    money_acc: int = 0


@dataclass
class SimState:
    t: float = 0.0
    pop: int = 100
    money: int = 9900
    T_stream: int = 0
    d_stream: int = 0
    ed01: int = 0
    ed02: int = 0
    ed03: int = 0
    ed04: int = 0
    ed05: int = 0
    ed51: int = 0
    talk_lv: int = 0
    movies: List[Movie] = field(default_factory=list)
    max_slots: int = 2
    # 累計（参考）
    total_money_in: int = 0
    total_views: int = 0

    def edit_mult(self) -> float:
        return (0.9**self.ed02)

    def ed04_mult(self) -> float:
        return 1.0 + 0.2 * self.ed04

    def stream_n(self) -> int:
        return max(0, self.T_stream - self.d_stream)

    def stream_sec(self) -> float:
        return streaming_final(self.stream_n(), self.d_stream)

    def edit_sec(self) -> float:
        return max(1.0, math.floor(120.0 * self.edit_mult()))

    def num_item_editors(self) -> int:
        # 0〜2m: MailChara が配信＋編集の連結パイプライン（ne=0）
        # 2m〜5m: 配信専念＋編集枠1（ItemEditor）
        # 5m〜: 編集枠2
        if self.t < 120:
            return 0
        if self.t < 300:
            return 1
        return 2

    def talk_mult(self) -> float:
        return 1.0 + 0.2 * min(self.talk_lv, 8)

    def spawn_movie_p(self) -> int:
        base = max(0, self.pop)
        adj = int(round(base * self.talk_mult()))
        return max(0, int(round(adj * 1.0 * self.ed04_mult())))

    def apply_purchases(self, minute: int, costs: List[Tuple[str, int]]):
        for name, price in costs:
            self.money -= price


def views_tick(m: Movie) -> Tuple[int, int]:
    """delta_views, popularity_delta (approx)"""
    p = max(0, m.p)
    delta = 0
    if not m.initial_done:
        initial = int(round(p * INITIAL_VIEWS_MULT))
        initial = max(1, initial)
        m.initial_done = True
        delta += initial

    now = m.elapsed
    nxt = now + VIEWS_TICK
    base_delta = max(0.0, target_views(nxt, p) - target_views(now, p))
    tick_raw = max(0.0, base_delta * stage_rand(nxt))
    m.pending_frac += tick_raw
    tick_delta = int(math.floor(m.pending_frac))
    if tick_delta > 0:
        m.pending_frac -= tick_delta
    delta += tick_delta

    dpop = int(round(delta * 0.2 * POP_RAND))
    m.elapsed = nxt
    m.tick_count += 1
    return delta, dpop


def money_subtick(m: Movie) -> int:
    m.pending_monetized += 0  # views already added in caller
    return 0


def run_sim(
    purchase_plan: List[Tuple[float, List[Tuple[str, int]]]],
    duration_s: float = 3600.0,
) -> SimState:
    st = SimState()
    st.movies.append(Movie(p=st.spawn_movie_p()))  # t=0 で1本投入
    views_accum = 0.0
    purchase_idx = 0

    # 次の動画投入時刻（パイプライン）
    def schedule_next_movie(from_t: float) -> float:
        s = st.stream_sec()
        e = st.edit_sec()
        ne = st.num_item_editors()
        if ne <= 0:
            return from_t + s + e
        # ストリーム1本あたり e 秒で1編集が消化、同時に ne 本並列
        # 近似: 周期 max(s, e/ne) は粗い。ここではボトルネック側を採用
        per_movie = max(s, e / ne)
        return from_t + per_movie

    next_spawn = schedule_next_movie(0.0)

    while st.t < duration_s + 1e-6:
        minute = int(st.t // 60)

        while purchase_idx < len(purchase_plan) and purchase_plan[purchase_idx][0] <= minute:
            _, costs = purchase_plan[purchase_idx]
            st.apply_purchases(minute, costs)
            purchase_idx += 1

        # スロット数更新
        st.max_slots = 2 + st.ed05

        # 新規映画
        while len(st.movies) < st.max_slots and st.t + 1e-6 >= next_spawn:
            st.movies.append(Movie(p=st.spawn_movie_p()))
            next_spawn = schedule_next_movie(next_spawn)

        dpop_total = 0
        for m in st.movies:
            dv, dp = views_tick(m)
            if dv > 0:
                st.total_views += dv
            dpop_total += dp
            views_accum += VIEWS_TICK
            # 10秒ごとの金銭: tick_count が 2 回で 10 秒相当（5秒×2）
            if m.tick_count % 2 == 0:
                payable = m.pending_monetized // 100
                if payable > 0:
                    pv = payable * 100
                    m.pending_monetized -= pv
                    rate = money_rate(m.p)
                    raw = pv * rate + m.pending_money_frac
                    dm = int(math.floor(max(0.0, raw)))
                    m.pending_money_frac = max(0.0, raw - dm)
                    st.money += dm
                    st.total_money_in += dm

        st.pop = max(0, st.pop + dpop_total)
        st.t += VIEWS_TICK

    return st


def build_purchase_plan_from_multipliers(
    ed02_bp, ed02_pm,
    ed03_bp, ed03_pm,
    ed04_bp, ed04_pm,
    ed05_bp, ed05_pm,
    ed51_bp, ed51_pm,
    st01_bp, st01_pm,
    st02_bp, st02_pm,
    st03_bp, st03_pm,
    st04_bp, st04_pm,
    ed01_price: int = 8000,
) -> List[Tuple[float, List[Tuple[str, int]]]]:
    """分単位でまとめた購入（同一分はリスト）。2回目は中点補間。"""
    def ed_lv_cost(bp, pm, lv):
        return ("ed", compute_sale_price(bp, pm, lv))

    def st_lv_cost(bp, pm, lv):
        return ("st", compute_sale_price(bp, pm, lv))

    # レベルは購入回数0始まり
    events: List[Tuple[float, str, int, Tuple]] = []

    def add_stair(tag, m1, m3, bp, pm, kind):
        m2 = (m1 + m3) / 2.0
        events.append((m1, tag, 0, (kind, bp, pm)))
        events.append((m2, tag, 1, (kind, bp, pm)))
        events.append((m3, tag, 2, (kind, bp, pm)))

    # EditorBuy は固定1000想定
    events.append((2.0, "eb01", 0, ("fixed", 1000)))
    events.append((10.0, "eb02", 0, ("fixed", 1000)))
    events.append((30.0, "eb03", 0, ("fixed", 1000)))

    events.append((5.0, "ed01", 0, ("ed01", ed01_price, 1.0)))

    add_stair("ed02", 7, 38, ed02_bp, ed02_pm, "ed02")
    add_stair("ed03", 8, 39, ed03_bp, ed03_pm, "ed03")
    add_stair("ed04", 9, 40, ed04_bp, ed04_pm, "ed04")
    add_stair("ed05", 11, 45, ed05_bp, ed05_pm, "ed05")

    events.append((25.0, "ed51", 0, ("ed51", ed51_bp, ed51_pm)))

    add_stair("st01", 4, 42, st01_bp, st01_pm, "st01")
    add_stair("st02", 8, 28, st02_bp, st02_pm, "st02")
    add_stair("st03", 9, 38, st03_bp, st03_pm, "st03")
    add_stair("st04", 10, 50, st04_bp, st04_pm, "st04")

    # 分ごとに集約
    from collections import defaultdict

    by_min: dict[int, List[Tuple[str, int]]] = defaultdict(list)
    st_levels = {"st01": 0, "st02": 0, "st03": 0, "st04": 0}
    ed_levels = {"ed02": 0, "ed03": 0, "ed04": 0, "ed05": 0}

    for tmin, _tag, _idx, payload in sorted(events, key=lambda x: x[0]):
        kind = payload[0]
        if kind == "fixed":
            by_min[int(tmin)].append(("editor", payload[1]))
        elif kind == "ed01":
            by_min[int(tmin)].append(("ed01", compute_sale_price(payload[1], payload[2], 0)))
        elif kind == "ed51":
            by_min[int(tmin)].append(("ed51", compute_sale_price(payload[1], payload[2], 0)))
        elif kind in ed_levels:
            lv = ed_levels[kind]
            by_min[int(tmin)].append((kind, compute_sale_price(payload[1], payload[2], lv)))
            ed_levels[kind] += 1
        elif kind in st_levels:
            lv = st_levels[kind]
            by_min[int(tmin)].append((kind, compute_sale_price(payload[1], payload[2], lv)))
            st_levels[kind] += 1

    plan: List[Tuple[float, List[Tuple[str, int]]]]] = []
    for mn in sorted(by_min.keys()):
        plan.append((float(mn), by_min[mn]))
    return plan


def apply_upgrade_side_effects(state: SimState, name: str):
    if name == "ed01":
        state.ed01 = 1
    elif name == "ed02":
        state.ed02 = min(3, state.ed02 + 1)
    elif name == "ed03":
        state.ed03 = min(5, state.ed03 + 1)
    elif name == "ed04":
        state.ed04 = min(5, state.ed04 + 1)
    elif name == "ed05":
        state.ed05 = min(3, state.ed05 + 1)
    elif name == "ed51":
        state.ed51 = 1
    elif name == "st01":
        state.T_stream += 1
        state.talk_lv = min(8, state.talk_lv + 1)
    elif name == "st02":
        state.T_stream += 1
    elif name == "st03":
        state.T_stream += 1
    elif name == "st04":
        state.T_stream += 1
        state.d_stream = min(5, state.d_stream + 1)


# --- 改良シミュ: 購入時にアップグレード状態を更新 ---
def run_sim_with_upgrades(plan: List[Tuple[float, List[Tuple[str, int]]]]) -> SimState:
    st = SimState()
    st.movies = [Movie(p=st.spawn_movie_p())]
    purchase_idx = 0

    def schedule_next_movie(from_t: float) -> float:
        s = st.stream_sec()
        e = st.edit_sec()
        ne = st.num_item_editors()
        if ne <= 0:
            return from_t + s + e
        per_movie = max(s, e / ne)
        return from_t + per_movie

    next_spawn = schedule_next_movie(0.0)

    while st.t < 3600.0 + 1e-6:
        minute = int(st.t // 60)

        while purchase_idx < len(plan) and plan[purchase_idx][0] <= minute:
            _, pairs = plan[purchase_idx]
            for name, price in pairs:
                st.money -= price
                if not name.startswith("editor"):
                    apply_upgrade_side_effects(st, name)
            purchase_idx += 1

        st.max_slots = 2 + st.ed05

        while len(st.movies) < st.max_slots and st.t + 1e-6 >= next_spawn:
            st.movies.append(Movie(p=st.spawn_movie_p()))
            next_spawn = schedule_next_movie(next_spawn)

        dpop_total = 0
        for m in st.movies:
            dv, dp = views_tick(m)
            if dv > 0:
                st.total_views += dv
                m.pending_monetized += dv
            dpop_total += dp
            if m.tick_count % 2 == 0:
                payable = m.pending_monetized // 100
                if payable > 0:
                    pv = payable * 100
                    m.pending_monetized -= pv
                    rate = money_rate(m.p)
                    raw = pv * rate + m.pending_money_frac
                    dm = int(math.floor(max(0.0, raw)))
                    m.pending_money_frac = max(0.0, raw - dm)
                    st.money += dm
                    st.total_money_in += dm

        st.pop = max(0, st.pop + dpop_total)
        st.t += VIEWS_TICK

    return st


if __name__ == "__main__":
    # 現行コード既定の倍率・基礎価格（St は const）
    plan = build_purchase_plan_from_multipliers(
        ed02_bp=50000,
        ed02_pm=4.0,
        ed03_bp=100000,
        ed03_pm=4.0,
        ed04_bp=1000,
        ed04_pm=4.0,
        ed05_bp=150000,
        ed05_pm=4.0,
        ed51_bp=200000,
        ed51_pm=1.0,
        st01_bp=1000,
        st01_pm=5.10210,
        st02_bp=3000,
        st02_pm=4.6,
        st03_bp=80000,
        st03_pm=8.2,
        st04_bp=320000,
        st04_pm=3.2,
        ed01_price=8000,
    )
    st = run_sim_with_upgrades(plan)
    print("pop_60m", st.pop, "money", st.money, "total_in", st.total_money_in, "views", st.total_views)
